src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,115 @@
 package core.basesyntax;
 
+import java.util.NoSuchElementException;
+import java.util.Objects;
+
 public class ArrayList<T> implements List<T> {
+    private static final int START_LENGTH = 10;
+    private T[] array;
+    private int size;
+
+    public ArrayList() {
+        array = (T[]) new Object[START_LENGTH];
+    }
+
     @Override
     public void add(T value) {
-
+        if (!enoughSpace()) {
+            grow();
+        }
+        array[size] = value;
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
-
+        if (!correctIndexForAdd(index)) {
+            throw new ArrayListIndexOutOfBoundsException("Cant add element at such index");
+        }
+        if (!enoughSpace()) {
+            grow();
+        }
+        if (index != size) {
+            for (int i = size - 1; i >= index; i--) {
+                array[i + 1] = array[i];
+            }
+        }
+        array[index] = value;
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        if (!correctIndexForGet(index)) {
+            throw new ArrayListIndexOutOfBoundsException("Cant get element with such index");
+        }
+        return array[index];
     }
 
     @Override
     public void set(T value, int index) {
-
+        if (!correctIndexForAdd(index)) {
+            throw new ArrayListIndexOutOfBoundsException("Cant set element with such index");
+        }
+        remove(index);
+        add(value, index);
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        if (!correctIndexForGet(index)) {
+            throw new ArrayListIndexOutOfBoundsException("Cant remove element with at index");
+        }
+        T removedValue = array[index];
+        for (int i = index; i < size - 1; i++) {
+            array[i] = array[i + 1];
+        }
+        size--;
+        return removedValue;
     }
 
     @Override
     public T remove(T element) {
-        return null;
+        for (int i = 0; i < size; i++) {
+            if (Objects.equals(array[i],(element))) {
+                return remove(i);
+            }
+        }
+        throw new NoSuchElementException("The element to delete was not found");
```

Comment:
```suggestion
        throw new NoSuchElementException("The element to delete was not found. Element " + element);
```

lets make exception messages more informative